### PR TITLE
A-matrix and inverse A-matrix

### LINKED ISSUE #859: A-matrix and inverse A-matrix
The A-matrix (AKA pedigree relatedness or the numerator relationship matrix) can be computed from pedigree kinship as kinship * ploidy for the fixed-ploidy case. The inverse of the A-matrix (**A**<sup>-1</sup>) is used in both pedigree-BLUP and single-step-GBLUP for genomic prediction. **A**<sup>-1</sup> can be computed by inverting **A** but this is impractical when **A** is large. Alternatively, there are efficient algorithms to calculate **A**<sup>-1</sup> directly from kinship values without inverting **A**. Hamilton and Kerr (2017) generalize this approach to the autopolyploid and mixed-ploidy cases.

One consideration is that **A**<sup>-1</sup> is a sparse matrix and doesn't strictly require computation of the *full* kinship matrix. Hence, we could potentially make use of [sparse](https://github.com/pydata/sparse/) (which is supported by dask). Though it may be simpler to start with dense matrices in the first instance.

### PR DESCRIPTION
**Pull request by timothymillar**
Fixes #859 and #860 

This implements calculation of the A-matrix and its inverse. Calculation of the inverse follows Hamilton and Kerr for both the diploid and autopolyploid cases. I don't think a diploid specific implementation is necessary for the inverse as calculation of the (non-zero) values is O(N).

Todo:

- [x] Docstring examples
- [x] Tests (diploid case and half-founders)

~~I'm also wondering whether the `pedigree_` prefix on these functions is useful for clarity or unnecessarily verbose?~~ **Edit** prefix removed

@tomwhite are you happy with the changes made to [`define_variable_if_absent`](https://github.com/pystatgen/sgkit/commit/bc2ad704cd6bdd7cc6ed6ce674c0ca17533e68ba)? I think the use of kwargs is justified in this case.

### CODE REVIEW

**Comment by timothymillar** [comment]
Looks like pre-commit is failing in CI due to some changes with mypy strictness?

**Comment by jeromekelleher** [comment]
Is it time to ditch mypy @tomwhite?

**Comment by timothymillar** [comment]
I think this is ready now, just waiting on #863 and #864 to be resolved so CI can pass.

**Comment by tomwhite** [comment]
Sorry this is blocked @timothymillar! I have merged the mypy change and opened #866 to hopefully get the build passing again.

**Comment by timothymillar** [comment]
No worries, there's no hurry from me!

**Comment by tomwhite** [comment]
@jeromekelleher are you able to review this one?

File: sgkit/stats/pedigree.py [hunk 1]
```diff
@@ -649,3 +649,368 @@ def pedigree_kinship(
...
+        mtx[q, i] -= frac
+        mtx[i, q] -= frac
+    if (p >= 0) and (q >= 0):
+        num = weight_p * weight_q * scalar
+        denom = 2 * np.sqrt(ploidy_p / 2 * ploidy_q / 2)
+        frac = num / denom
+        mtx[p, q] += frac
+        mtx[q, p] += frac
+    mtx[i, i] += scalar / ploidy_i
+    return
```
Comment (jeromekelleher) [inline sgkit/stats/pedigree.py:844]: I find the empty return odd Pythonically, but perhaps there's a good reason from a jitting perspective?

**Review by jeromekelleher — APPROVED** [review]
Awesome stuff @timothymillar :muscle: 

One question, please add the "automerge" label when you're happy to merge.

**Comment by timothymillar** [comment]
Just noting that `sgkit/tests/test_stats_utils.py::test_concat_2d__variables` failed on win_build 3.8 for comit [6bd6b96](https://github.com/pystatgen/sgkit/pull/862/commits/6bd6b96cac509d8ffde8e4ac6b22b5500cffd34d)  with the message:

"Unreliable test timings! On an initial run, this test took 279.38ms, which exceeded the deadline of 200.00ms, but on a subsequent run it took 0.92 ms, which did not. If you expect this sort of variability in your test timings, consider turning deadlines off for this test by setting deadline=None."

I'm assuming that this is just an outlier as I haven't touched this code so will restart that run.

**Pull request merged**